executor/oci/spec_unix.go

```diff
@@ -4,77 +4,45 @@ package oci
 	"github.com/containerd/containerd/oci"
-	"github.com/containerd/continuity/fs"
 	"github.com/docker/docker/pkg/idtools"
-	"github.com/mitchellh/hashstructure"
-	"github.com/moby/buildkit/executor"
-	"github.com/moby/buildkit/snapshot"
 	"github.com/moby/buildkit/solver/pb"
 	"github.com/moby/buildkit/util/entitlements/security"
-	"github.com/moby/buildkit/util/network"
 	"github.com/moby/buildkit/util/system"
 	specs "github.com/opencontainers/runtime-spec/specs-go"
-	"github.com/pkg/errors"
 )
 
-// Ideally we don't have to import whole containerd just for the default spec
-
-// GenerateSpec generates spec using containerd functionality.
-// opts are ignored for s.Process, s.Hostname, and s.Mounts .
-func GenerateSpec(ctx context.Context, meta executor.Meta, mounts []executor.Mount, id, resolvConf, hostsFile string, namespace network.Namespace, processMode ProcessMode, idmap *idtools.IdentityMapping, opts ...oci.SpecOpts) (*specs.Spec, func(), error) {
-	c := &containers.Container{
-		ID: id,
-	}
-	_, ok := namespaces.Namespace(ctx)
-	if !ok {
-		ctx = namespaces.WithNamespace(ctx, "buildkit")
-	}
-	if meta.SecurityMode == pb.SecurityMode_INSECURE {
-		opts = append(opts, security.WithInsecureSpec())
-	} else if system.SeccompSupported() && meta.SecurityMode == pb.SecurityMode_SANDBOX {
-		opts = append(opts, seccomp.WithDefaultProfile())
+func generateSecurityOpts(mode pb.SecurityMode) ([]oci.SpecOpts, error) {
+	if mode == pb.SecurityMode_INSECURE {
+		return []oci.SpecOpts{security.WithInsecureSpec()}, nil
+	} else if system.SeccompSupported() && mode == pb.SecurityMode_SANDBOX {
+		return []oci.SpecOpts{seccomp.WithDefaultProfile()}, nil
 	}
+	return nil, nil
+}
 
-	switch processMode {
-	case NoProcessSandbox:
-		// Mount for /proc is replaced in GetMounts()
-		opts = append(opts,
-			oci.WithHostNamespace(specs.PIDNamespace))
+func generateProcessModeOpts(mode ProcessMode) ([]oci.SpecOpts, error) {
+	if mode == NoProcessSandbox {
+		// Mount for /proc is replaced in GetMounts() anyway
+		return []oci.SpecOpts{oci.WithHostNamespace(specs.PIDNamespace)}, nil
 		// TODO(AkihiroSuda): Configure seccomp to disable ptrace (and prctl?) explicitly
 	}
+	return nil, nil
+}
 
-	// Note that containerd.GenerateSpec is namespaced so as to make
-	// specs.Linux.CgroupsPath namespaced
-	s, err := oci.GenerateSpec(ctx, nil, c, opts...)
-	if err != nil {
-		return nil, nil, err
-	}
-	// set the networking information on the spec
-	namespace.Set(s)
-
-	s.Process.Args = meta.Args
-	s.Process.Env = meta.Env
-	s.Process.Cwd = meta.Cwd
-	s.Process.Rlimits = nil           // reset open files limit
-	s.Process.NoNewPrivileges = false // reset nonewprivileges
-	s.Hostname = "buildkitsandbox"
-
-	s.Mounts, err = GetMounts(ctx,
+func setupForLinux(ctx context.Context, c *containers.Container, s *specs.Spec, securityMode pb.SecurityMode, resolvConf, hostsFile string, processMode ProcessMode, idmap *idtools.IdentityMapping) error {
```

Comment:
Can we split this into more use-case based functions and avoid passing 10 parameters to a function. `setupMaskedPaths` `setupIDMapping` `markInsecure` etc.

Reply:
It might be better to refactor it so that the actual intended behaviours fit together. Right now a lot of these are operating on mounts anyway, so it'd be clearer to just build the mounts one-at-a-time, taking all the things that affect them at once, I think, rather than generating a bunch and then munging or deleting them one-by-one.

I had been considering refactoring this function to generate a whole bunch of `SpecOpts` rather than a few `SpecOpts` passed to `GenerateSpec` and then post-processing with the same flags, including more `SpecOpts` hand-applied. I'll have a look, that might be cleaner to read anyway.

Reply:
To do this refactoring, I'm going to replace the top commit (923a995), because it makes a lot more sense to do the refactoring from 959d0eb.